source/win32/low_latency_iocp_context.cpp

```diff
@@ -0,0 +1,682 @@
+/*
+ * Copyright 2020-present Facebook, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#pragma once
+
+#include <unifex/win32/low_latency_iocp_context.hpp>
+#include <unifex/scope_guard.hpp>
+
+#include <atomic>
+#include <system_error>
+#include <random>
+
+#define WIN32_LEAN_AND_MEAN
+#include <Windows.h>
+
+namespace unifex::win32 {
+
+namespace {
+
+static safe_handle create_iocp() {
+    HANDLE h = ::CreateIoCompletionPort(INVALID_HANDLE_VALUE, nullptr, 0, 1);
+    if (h == NULL) {
+        DWORD errorCode = ::GetLastError();
+        throw std::system_error{static_cast<int>(errorCode), std::system_category(), "CreateIoCompletionPort()"};
+    }
+
+    return safe_handle{h};
+}
+
+} // namespace
+
+low_latency_iocp_context::low_latency_iocp_context(std::size_t maxIoOperations)
+: activeThreadId_(std::thread::id())
+, iocp_(create_iocp())
+, ioPoolSize_(maxIoOperations)
+, ioPool_(std::make_unique<vectored_io_state[]>(maxIoOperations)) {
+
+    // Build the I/O free-list in reverse so front of free-list
+    // is first element in array.
+    for (std::size_t i = 0; i < maxIoOperations; ++i) {
+        ioFreeList_.push_front(&ioPool_[maxIoOperations - 1 - i]);
+    }
+}
+
+low_latency_iocp_context::~low_latency_iocp_context() {
+    // Wait until the completion-event for every io-state has been 
+    // received before we free the memory for the io-states.
+    std::size_t remaining = ioPoolSize_;
+    while (!ioFreeList_.empty()) {
+        (void)ioFreeList_.pop_front();
+        --remaining;
+    }
+
+    if (remaining > 0) {
+        constexpr std::uint32_t completionBufferSize = 128;
+        OVERLAPPED_ENTRY completionBuffer[completionBufferSize];
+        std::memset(&completionBuffer, 0, sizeof(completionBuffer));
+
+        do {
+            DWORD numEntriesRemoved = 0;
+            BOOL ok = ::GetQueuedCompletionStatusEx(
```

Comment:
Be aware that this batch function is umpteen times slower than the single packet edition. It makes sense if you are pulling down many completions at a time, it amortises out, but if you know that you're not, looping processing them singly can be very considerably faster.

Reply:
So should I be keeping track of the number of pending I/Os and only calling the batch API if this number is above some threshold?

I can imagine cases where, even though I have a reasonable number of outstanding I/Os, that I might still only have one at a time coming in.

What heuristics should this code be using to determine whether to call the single vs batch API?

Reply:
In my first design, I kept a count, and if outstanding i/o were below 16 or so, I looped the single-completion API.

In my second design, I eliminated the count, and made when completions get done parameterised. I now implement completions by doing completions depending on context, so instead of there being a global `run()`, they get dispatched ad-hoc, wherever they need to be. As the caller knows whether they need to spin fast or don't care, they can dictate whether to complete one, two, or as many as possible completions. I found this design much better. But I appreciate that that isn't an Executor, and so it probably doesn't help you.